### PR TITLE
Nick first pass at porting to python

### PR DESCRIPTION
**Pull request by NickCEBM**
First try at porting the data pipeline from SQL to Python.

Re: https://github.com/ebmdatalab/clinicaltrials-act-tracker/issues/181

### CODE REVIEW

**Comment by NickCEBM** [comment]
For the test files, this creates an exact replica. Should test on the full dataset and ensure all the numbers currently match up.

A few minor improvements discovered during the porting process can be integrated.

File: nick_work/fdaaa_port.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            if cd is not None:
+                try:
+                    completion_date = datetime.strptime(cd.text, '%B %d, %Y').date()
+                    defaulted_cd = 0
+                except ValueError:
+                    completion_date = datetime.strptime(cd.text, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
+                    defaulted_cd = 1
+            else:
+                completion_date = None
+                defaulted_cd = 0
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: The above three date operations involve a lot of copy-and-paste, which indicates another function should probably be used.

Should also include comments to explain the relativedelta choice.
Something like this:

```python
def string_to_date(datestr):
    """Convert text to date. If the day is missing, pick the last day of the current month.
    Returns (date, is_defaulted_date) tuple.
    """
    is_defaulted_date = False
    try:
        parsed_date = datetime.strptime(datestr, '%B %d, %Y').date()
    except ValueError:
        parsed_date = datetime.strptime(datestr, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
        is_defaulted_date = True
    return (is_defaulted_date, parsed_date)

is_defaulted_date, completion_date = string_to_date(soup.completion_date)
```

File: nick_work/fdaaa_port.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+    for file in os.listdir(directory):
+        if file.endswith('.xml'):
+            xml_bs4 = open(directory + '/' + file)
+            xml_contents = xml_bs4.read()
+            soup = BeautifulSoup(xml_contents,'xml')
+            xml_json = open(directory + '/' + file, 'rb')
+            parsed_json = xmltodict.parse(xml_json.read())
+
+            #nct_id
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: These comments are redundant as they add nothing beyond repeating the variable assignment.

File: nick_work/fdaaa_port.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                available_completion_date = None
+            elif primary_completion_date == None and completion_date is not None:
+                available_completion_date = completion_date
+                used_cd = 1
+                used_pcd = 0
+            else:
+                available_completion_date = primary_completion_date
+                used_pcd = 1
+                used_cd = 0
+
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: This would be more readable:

```python
            if not primary_completion_date and not completion_date:
                available_completion_date = None
            elif completion date and not primary_completion_date:
                available_completion_date = completion_date
                used_cd = True
                used_pcd = False
```

File: nick_work/fdaaa_port.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+            #act_flag - #looks like != handles nulls in Python so no need for extra term we use in BQ
+            if study_type == 'Interventional' and\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: where an expression is in parentheses you don't need to escape newlines. So I'd add parens around the whole lot...

File: nick_work/fdaaa_port.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+            (fda_reg_drug == 'Yes' or fda_reg_device == 'Yes') and\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: ...and you didn't need the newline escape here (and below) anyway because already in parens

File: nick_work/fdaaa_port.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            (
+             phase == 'Phase 1/Phase 2' or phase == 'Phase 2' or\
+             phase == 'Phase 2/Phase 3' or phase == 'Phase 3' or\
+             phase == 'Phase 4' or phase == 'N/A'
+            ) and\
+            primary_purpose != 'Device Feasibility' and\
+            start_date >= effective_date and\
+            study_status != 'Withdrawn':
+                act_flag = 1
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Prefer `True` and `False` again (and throughout)

File: nick_work/fdaaa_port.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+            else:
+                act_flag = 0
+
+            #old_pact_flag
+            for it in soup.find_all("intervention_type"):
+                if ('Drug' or 'Device' or 'Biological' or 'Genetic' or\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Same point about escapes

File: nick_work/fdaaa_port.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+                used_cd = 0
+
+            #act_flag - #looks like != handles nulls in Python so no need for extra term we use in BQ
+            if study_type == 'Interventional' and\
+            (fda_reg_drug == 'Yes' or fda_reg_device == 'Yes') and\
+            (
+             phase == 'Phase 1/Phase 2' or phase == 'Phase 2' or\
+             phase == 'Phase 2/Phase 3' or phase == 'Phase 3' or\
+             phase == 'Phase 4' or phase == 'N/A'
+            ) and\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: We use this expression repeatedly. Probably worth making a method like

```python
def phase_is_act_compatible(phase):
    return phase in ['Phase 1/Phase 2', 'Phase 2', 'Phase 2/Phase 3' , 'Phase 3', 'Phase 4', 'N/A']
```

File: nick_work/fdaaa_port.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+            primary_purpose != 'Device Feasibility' and\
+            start_date >= effective_date and\
+            study_status != 'Withdrawn':
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Previous point also potentially applies (I think) to these.  i.e. factor these out into a descriptive function for readability.

File: nick_work/fdaaa_port.py (hunk 3 above)
Comment (sebbacon) [reply nick_work/fdaaa_port.py]: Also: you rename `used_pcd` to `used_primary_completion_date` later. You might as well just call it `used_primary_completion_date` and save a line of code (same goes for `used_cd`)

File: nick_work/fdaaa_port.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                title = official_title
+            elif official_title is None and brief_title is not None:
+                title = brief_title
+            else:
+                title = None
+
+            #discrep_date_status - add this to "if" statement after testing: 'and completion_date is not null')
+            if (primary_completion_date < date.today() or primary_completion_date is None) and\
+            completion_date < date.today() and\
+            study_status is ("Unknown status" or "Active, not recruiting" or "Not yet recruiting" or\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Don't need this newline escape

File: nick_work/fdaaa_port.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+            #title
+            if official_title is not None:
+                title = official_title
+            elif official_title is None and brief_title is not None:
+                title = brief_title
+            else:
+                title = None
+
+            #discrep_date_status - add this to "if" statement after testing: 'and completion_date is not null')
+            if (primary_completion_date < date.today() or primary_completion_date is None) and\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Add parens around the whole lot and avoid ugly escapes

File: nick_work/fdaaa_port.py [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+            except KeyError:
+                intervention_mesh = None
+
+            #keywords
+            try:
+                keywords = json.dumps(parsed_json['clinical_study']['keyword'])
+            except KeyError:
+                keywords = None
+
+            row_list = [nct_id, act_flag, included_pact_flag, has_results, pending_results, pending_data, has_certificate,\
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Redundant backslash

File: nick_work/fdaaa_port.py [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+
+            #defaulted_date
+            if (used_primary_completion_date == 1 and defaulted_pcd_flag == 1) or\
+            used_primary_completion_date == 0 and defaulted_cd_flag == 1:
+                defaulted_date = 1
+            else:
+                defaulted_date = 0
+
+            #condition
+            try:
```
Comment (sebbacon) [inline nick_work/fdaaa_port.py]: Make a little function to avoid repeating identical try/excepts here and below: `json_or_none()` or similar.

File: nick_work/fdaaa_port.py [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+from datetime import datetime
+from datetime import timedelta
+from dateutil.relativedelta import relativedelta
+import pandas as pd
+import csv
+cs = 'clinical_study'
+
+projectid = "ebmdatalab"
+
+fda_reg_df = pd.read_gbq('''
```
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Yes, we can make this just a static import of the file. Not sure how we get around using Pandas to work with it though per your comment below.

File: nick_work/fdaaa_port.py (hunk 2 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Yes, these were just for ease of arranging as I went along to make sure I had everything. I can remove for the final version.

File: nick_work/fdaaa_port.py [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+           'intervention_mesh', 'keywords']
+
+effective_date = date(2017, 1, 18)
+directory = 'NCTxxx' #this points to a folder on my working directory. Change as needed.
+
+with open('test_output.csv', 'w') as test_csv:
+    writer = csv.writer(test_csv)
+    writer.writerow(headers)
+
+    for file in os.listdir(directory):
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: "xml_filename" instead of clobbering function name "file"

File: nick_work/fdaaa_port.py (hunk 4 above)
Comment (chadmiller) [reply nick_work/fdaaa_port.py]: ...or nest some ifs, if you're already multiline.  Makes it more readable, up to a point.

```
if study_type == "Interventional":
    if fda_reg_drug == "Yes" or fda_reg_device == "Yes":
        if phase in (...)
```

File: nick_work/fdaaa_port.py [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+import xmltodict
+import json
+from datetime import date
+from datetime import datetime
+from datetime import timedelta
+from dateutil.relativedelta import relativedelta
+import pandas as pd
+import csv
+cs = 'clinical_study'
+
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]:
```
def t(textish):
    if textish is None:
        return None
    return textish.text
```

Use this every .text .

File: nick_work/fdaaa_port.py [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+            if hc is not None:
+                has_certificate = 1
+            else:
+                has_certificate = 0
+
+            #phase
+            phase = (soup.phase).text    
+
+            #fda_reg_drug
+            if soup.is_fda_regulated_drug is None:
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: These four lines replaced with 
`fda_reg_drug = t(soup.is_fda_regulated_drug)`

File: nick_work/fdaaa_port.py (hunk 10 above)
Comment (chadmiller) [reply nick_work/fdaaa_port.py]: Those "or"s are wrong, and so is the "is".

(study status is the same exact object as True.)

`if study_status in (...):`

File: nick_work/fdaaa_port.py (hunk 12 above)
Comment (chadmiller) [reply nick_work/fdaaa_port.py]: And these are risky because you can have a value be initialized in a previous row and not set in this one, and you wouldn't catch it.

You would do better to init a dictionary at the top and think of it as your namespace, and fill it  (`v["location"] = ...`), and down here, write the dictionary. CSV module handles dictionaries. `writer.writedict(v)`.

This is also dangerous because you're assuning the same order here as in the header list above, which will change over time and eventually break. A dictionary is better.

File: nick_work/fdaaa_port.py [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+
+            #study_type
+            study_type = (soup.study_type).text
+
+            #has_certificate
+            hc = soup.disposition_first_submitted
+            if hc is not None:
+                has_certificate = 1
+            else:
+                has_certificate = 0
```
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Done. All dummy variable are booleans now.

File: nick_work/fdaaa_port.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Figured out how to do this. Just made a dictionary.

File: nick_work/fdaaa_port.py [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+                primary_purpose = soup.find('primary_purpose').text
+            else:
+                primary_purpose = None
+
+            #is_fda_regulated
+            df_row = fda_reg_df.loc[fda_reg_df['nct_id'] == nct_id]
+            if df_row.empty:
+                is_fda_regulated = None
+            else:
+                is_fda_regulated = df_row['is_fda_regulated'].values[0]
```
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Dumped pandas.

File: nick_work/fdaaa_port.py (hunk 1 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: done

File: nick_work/fdaaa_port.py (hunk 2 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: These are now removed throughout.

File: nick_work/fdaaa_port.py (hunk 3 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Done.

File: nick_work/fdaaa_port.py (hunk 4 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: I think the multi-line conditional is much easier to read than the nested ifs, especially now that it is much cleaner with the functions. Have done it that way for now and removed all line breaks throughout in favor of parens.

File: nick_work/fdaaa_port.py (hunk 6 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Done.

File: nick_work/fdaaa_port.py (hunk 8 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Created a whole host of functions now.

File: nick_work/fdaaa_port.py (hunk 10 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Thanks! Fixed both new lines and the error in the code.

File: nick_work/fdaaa_port.py (hunk 11 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Done.

File: nick_work/fdaaa_port.py (hunk 12 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Thanks for the tip Chad. I might not have implemented this the way you had in mind and this might not be the best way, but weather than using `writer.writedict()` what I did instead was create a dictionary that has the eventual CSV columns as the keys, and the values are lists containing the values for each trial. I then create a header from the key values and my rows from the lists.

Let me know if there will be issues with this at scale!

File: nick_work/fdaaa_port.py (hunk 9 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Done. Made all the functions.

File: nick_work/fdaaa_port.py (hunk 15 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Thanks!

File: nick_work/fdaaa_port.py [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+effective_date = date(2017, 1, 18)
+directory = 'NCTxxx' #this points to a folder on my working directory. Change as needed.
+
+with open('test_output.csv', 'w') as test_csv:
+    writer = csv.writer(test_csv)
+    writer.writerow(headers)
+
+    for file in os.listdir(directory):
+        if file.endswith('.xml'):
+            xml_bs4 = open(directory + '/' + file)
```
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Great tips, thanks, Implemented. My first run at this I had not yet discovered `os.path.join()` but have since (and changed a whole bunch of other scripts I have) and just hadn't gotten around to updating this one yet!

File: nick_work/fdaaa_port.py (hunk 16 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Implemented throughout. Thanks!.

**Comment by NickCEBM** [comment]
Updated file has been committed. Ignore the bit about the `dict_builder` function as I ended up just moving them all to the bottom.

File: nick_work/fdaaa_port.py [hunk 21]
```diff
@@ -6,367 +6,310 @@
...
+
+def dict_builder(your_dict,name,value):
+    your_dict[name].append(value)
+    return your_dict
 
 effective_date = date(2017, 1, 18)
 directory = 'NCTxxx' #this points to a folder on my working directory. Change as needed.
 
-with open('test_output.csv', 'w') as test_csv:
+trial_dict = defaultdict(list)
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: I want you to twist your table 90 degrees here.

Instead of making a dictionary containing lists, you should have made a list of dictionaries. Organize by a bunch of rows, not by a bunch of columns. (The danger of dict-of-lists is that you are later hoping that item \#574243 in list "foo" goes with the item \#574243 in list "bar", and that is super fragile. A later change where a loop over XML things doesn't insert a "foo" but does insert a "bar", and everything is wrong for the remaining ones.)

And in fact, I want to throw away the idea of holding many studies in memory at a time. There is no need to. Construct one, then write it out and move on.

So, when you read in a XML file and parse it, make a new, single dictionary there. Later, hand that dictionary to the CSV writer before you go to the next study.

Catch all values in the dict, instead of local variables.

```
    s["phase"] = t(soup.phase)
```

File: nick_work/fdaaa_port.py [hunk 22]
```diff
@@ -6,367 +6,310 @@
...
-            row_list = [nct_id, act_flag, included_pact_flag, has_results, pending_results, pending_data, has_certificate,\
-                       results_due, start_date, available_completion_date, used_primary_completion_date, defaulted_pcd_flag,\
-                       defaulted_cd_flag, results_submitted_date, last_updated_date, certificate_date, phase, enrollment,\
-                       location, study_status, study_type, primary_purpose, sponsor, sponsor_type, collaborators, exported,\
-                       fda_reg_drug, fda_reg_device, is_fda_regulated, url, title, official_title, brief_title,\
-                       discrep_date_status, late_cert, defaulted_date, condition, condition_mesh, intervention, intervention_mesh,\
-                       keywords]
-
-            writer.writerow(row_list)
+    writer.writerows(zip(*trial_dict.values()))
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: Whoa, the results of dict.values is not guaranteed to be in any order (in some versions of Python). You shouldn't disassociate the keys from values. `csv` module manages the hard part.

```python
ordered_csv_column_names = ('nct_id', 'act_flag', 'included_pact_flag', ...)

with open('test_output.csv', 'w', newline='',encoding='utf-8') as test_csv:
    csv_file = csv.DictWriter(test_raw_file, fieldnames=ordered_csv_column_names)
    csv_file.writeheader()

    for xml in ...
        s = dict()
        s["phase"] = ...

        csv.file.writerow(s)
```

File: nick_work/fdaaa_port.py [hunk 23]
```diff
@@ -6,367 +6,310 @@
...
+
+def is_not_withdrawn(study_status):
+    return study_status != 'Withdrawn'
+
+def is_interventional(study_type):
+    return study_type == 'Interventional'
+
+def is_covered_intervention(intervention_type):
+    for it in intervention_type:
+        if ('Drug' or 'Device' or 'Biological' or 'Genetic' or
```
[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: So, I sort of understand what you are saying, but I'm also a bit confused because I'm getting the right answer out the other end of the spreadsheet, where I would not be if this function wasn't operating as intended.

Either way, I understand how this approach works and as I am but a beginner I will absolutely be taking your word that this way is better. I've made this change to both this function and the `has_us_loc` function. The one difference is I added a `break` after finding a single `True` as in this instance, I don't care how many of the covered interventions or locations are listed as part of the trial, I just care about the first instance of any of them.

[the remark this answers is not in the document]
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Alright, so I was just testing this all out and this method breaks defining pACTs that rely on these two fields while my old method works BUT I think that's just based on some fluky luck with all the interventions I'm testing being Drugs and all the location countries being just United States. After some fiddling, I think I figured out the mechanics of all this and how to fix it so it will work in all instances:

What is going on here is that
`intervention_type = soup.find_all("intervention_type")`
is returning a list of bs4 tags that look like this:
`[<intervention_type>Drug</intervention_type>, <intervention_type>Drug</intervention_type>]`

Originally, we were looking for a string within those tags. That's a bit clunky so I'm adding this to turn just the text of each intervention type into a list.

```python
intervention_type_field = soup.find_all("intervention_type")
    trial_intervention_types = []
    for tag in intervention_type_field:
        trial_intervention_types.append(tag.get_text())
```

Then with some googling, I found a solution for comparing common elements in two lists with sets that I think is quite simple and elegant for this application, so I think I'm going to implement that:

```Python
def is_covered_intervention(intervention_type_list):
    covered_intervention_type = ['Drug','Device','Biological','Genetic','Radiation','Combination Prodcut','Diagnostic Test']
    a_set = set(covered_intervention_type)
    b_set = set(intervention_type_list)
    if (a_set & b_set):
        return True
    else:
        return False
```

Getting a US location is implemented differently as:

`locs = t(soup.location_countries)`

just gives you a big string of all the countries involved. Our test dataset only has trials with a single US location, so I went and downloaded some additional trial XMLs with multiple countries and played around with those to confirm this. It is indeed just one enormous string with all the country names separated by line breaks.

So, this is actually a bit easier to handle. We can just look for the strings of the US locations inside of that string:

```python
def has_us_loc(locs):
    us_locs = ['United States','American Samoa','Guam','Northern Mariana Islands','Puerto Rico','Virgin Islands (U.S.)']
    for us_loc in us_locs:
        if us_loc in locs:
            return True
            break
        else:
            return False
```

I've just made my included intervention types and country strings their own lists within the functions as I think this is more readable.

File: nick_work/fdaaa_port.py (hunk 21 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: This makes sense and definitely easier to implement! The one thing to note is that some variables both go into the resulting CSV and are used again within the expressions or functions that determine other variables so I think in those instances, I'll have to both set the variable and assign it to the dictionary in 2 separate lines.

File: nick_work/fdaaa_port.py (hunk 22 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Yes, I knew this but thought for our purposes it wouldn't matter, but that's not good practice!

I've implemented your idea here all the way through and that fixes this problem!

File: nick_work/fdaaa_port.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (chadmiller) [reply nick_work/fdaaa_port.py]: When some "and" or "or" expression is evaluated, instead of making a new true or false value, Python uses the value that made the expression true or false. See it in the interactive interpreter (which is amazingly handy):


```
$ python3

Python 3.5.2 (default, Nov 12 2018, 13:43:14) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> ('Drug' or 'Device' or 'Biological' or 'Genetic' or 'Radiation' or 'Combination Prodcut' or 'Diagnostic Test')
'Drug'

>>> 'trueness' or ''
'trueness'

>>> 'trueness' and ''
''
```

If it was working, it must be that every study is a Drug study. I wanted you to see the operations and their orders. The first evaluation is list of ORs

`if ('Drug' or 'Device' or 'Biological' or 'Genetic' or 'Radiation' or 'Combination Prodcut' or 'Diagnostic Test') in it`

next becomes

`if ('Drug') in it`

which next becomes (because the parens around a simple expression is just that expression)

`if 'Drug' in it`

which next (depending on `it`) becomes

`if True`

which Python knows what to do with.

File: nick_work/fdaaa_port.py (hunk 21 above)
Comment (chadmiller) [reply nick_work/fdaaa_port.py]: Er, maybe I don't understand, but dictionaries aren't write-only. You can use `s['thingpreviouslyset"]` to read as well as write. I guess you might be mutating it into something else you don't want also to store, so in some cases you might want an extra variable.

File: nick_work/fdaaa_port.py [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+
+    sponsor = t(soup.sponsors.lead_sponsor.agency)
+
+    sponsor_type = t(soup.sponsors.lead_sponsor.agency_class)
+
+    collaborators = dict_or_none(parsed_json, [cs, 'sponsors', 'collaborator'])  
+
+    exported = t(soup.oversight_info.is_us_export)
+        
+    url = (soup.url).text
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: t() ?

File: nick_work/fdaaa_port.py [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+    official_title = t(soup.official_title)
+
+    brief_title = t(soup.brief_title)
+
+    if official_title is not None:
+        title = official_title
+    elif official_title is None and brief_title is not None:
+        title = brief_title
+    else:
+        title = None
```
Comment (chadmiller) [inline nick_work/fdaaa_port.py]: Remember the `or` lesson of how it yields a component of the items instead of True or False?

`title = official_title or brief_title`

File: nick_work/fdaaa_port.py (hunk 24 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Yep.

File: nick_work/fdaaa_port.py (hunk 25 above)
Comment (NickCEBM) [reply nick_work/fdaaa_port.py]: Ah, that's really nifty to know!

**Comment by sebbacon** [comment]
Closing for #195
